graders/python/python_autograder/code_feedback.py

```diff
@@ -299,6 +299,112 @@ def bad(msg):
+        - ``report_failure``: If true, feedback will be given on failure.
+        """
+
+        def bad(msg):
+            if report_failure:
+                cls.add_feedback(msg)
+            if accuracy_critical:
+                cls.finish("")
+            else:
+                return False
+
+        if data is None:
+            return bad(f"{name} is None or not defined")
+
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        if partial_keys is not None and len(partial_keys) >= 1:
+            for partial_key in partial_keys:
+                if partial_key not in data:
+                    return bad(f"{name} does not contain key {partial_key}")
+            return True
+
+        if partial_keys is None:
+            if check_only_values:
+                if len(ref.values()) != len(data.values()):
+                    return bad(
+                        f"{name} has the wrong length for values--expected {len(ref.values())}, got {len(data.values())}"
+                    )
+
+            if len(ref) != len(data):  # this is default length of keys check
+                return bad(
+                    f"{name} has the wrong number of entries, expected {len(ref)}, got {len(data)}"
+                )
+
+        if entry_type_value is not None:
+            for value in data.values():
+                if not isinstance(value, entry_type_value):
+                    return bad(
+                        f"{name} has the wrong type for value {value}, expecting type {entry_type_value}"
+                    )
+
+        if entry_type_key is not None:
+            for key in data.keys():
+                if not isinstance(key, entry_type_key):
+                    return bad(
+                        f"{name} has the wrong type for key {key}, expecting type {entry_type_key}"
+                    )
+
+        if check_only_keys or check_only_values:
+            check_keys = False
+            if check_only_keys:
+                for key in data.keys():
+                    if key not in ref.keys():
+                        return bad(f"{name} contains an extra key: {key}")
+                check_keys = True
+
+            check_values = False
+            if check_only_values:
+                if len(ref.values()) != len(data.values()):
+                    return f"{name} has the wrong length for values--expected {len(ref.values())}, got {len(data.values())}"
+                for value in data.values():
+                    if value not in ref.values():
+                        return bad(f"{name} contains an extra value: {value}")
+                check_values = True
+
+            if check_only_keys and check_only_values:
+                return check_keys and check_values
+
+            return check_keys or check_values
+
+        if (
+            ref == data
+        ):  # will check equality of both keys and values between reference dict and student's dict
+            return True
```

Comment:
Nit for better formatting:

```suggestion
        # Check equality of both keys and values between reference dict and student's dict
        if ref == data:  
            return True
```